src/background/messenger.ts

```diff
@@ -19,36 +20,58 @@ export interface ExtensionAdapter {
 }
 
 export default class Messenger {
-    private reporters: Set<(info: ExtensionData) => void>;
     private adapter: ExtensionAdapter;
+    private changeListenerCount: number;
 
     constructor(adapter: ExtensionAdapter) {
-        this.reporters = new Set();
         this.adapter = adapter;
-        chrome.runtime.onConnect.addListener((port) => {
-            if (port.name === 'ui') {
-                port.onMessage.addListener(async (message) => await this.onUIMessage(port, message));
+        this.changeListenerCount = 0;
+        chrome.runtime.onMessage.addListener((message: Message, _: any, sendResponse: (any) => void) => {
+            if (message.name === 'ui') {
+                this.onUIMessage(message, sendResponse);
                 this.adapter.onPopupOpen();
+                return ([
+                    'get-data',
+                    'get-active-tab-info'
+                ].includes(message.type));
             }
         });
+
+        // This is a work-around for Firefox bug which does not let to responding to onMessage handler above.
+        if (isFirefox) {
+            chrome.runtime.onConnect.addListener((port) => {
+                let promise = null;
```

Comment:
`promise` will be override anyway, not really a need to initialize it, but let's add a type definition.
```suggestion
                let promise: Promise<(data) => void>;
```

Reply:
I don't think `Promise<(data) => void>` is correct. Since this promise takes on value `Promise<ExtensionData>` or `Promise<TabInfo>`, I changed it to `Promise<ExtensionData | TabInfo>`. (It compiles and works as expected.)

Reply:
👍🏽